test/e2e/framework/certificates/certificates.go

```diff
@@ -0,0 +1,242 @@
+package certificates
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"strings"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	certType       = "CERTIFICATE"
+	privateKeyType = "PRIVATE KEY"
+)
+
+// ValidateCert checks if the certificate imported is valid for the
+// given dns name
+func ValidateCert(certData, dnsName string) {
+	roots := x509.NewCertPool()
+	// Without passing a pool, Go will use the system pool which will definitely not work.
+	// By adding the certificate itself, a valid path can be built to a trusted root for self-signed certs.
+	// Ref: https://stackoverflow.com/questions/63317763/check-validity-of-ssl-self-signed-certificate
+	ok := roots.AppendCertsFromPEM([]byte(certData))
+	Expect(ok).To(BeTrue())
+
+	block, _ := pem.Decode([]byte(certData))
+	cert, err := x509.ParseCertificate(block.Bytes)
+	Expect(err).To(BeNil())
+
+	opts := x509.VerifyOptions{
+		DNSName: dnsName,
+		Roots:   roots,
+	}
+
+	_, err = cert.Verify(opts)
+	By(fmt.Sprintf("Ensuring certificate is valid for dns name %s", dnsName))
+	Expect(err).To(BeNil())
+}
+
+// ValidateCertAndPrivateKey validates the certificate and private key returned by the provider match
+// and are usable
+func ValidateCertAndPrivateKey(data, dnsName string) {
+	By(fmt.Sprintf("Ensuring certificate and private key is valid for dns name %s", dnsName))
+	certPEMBlock, err := getCert([]byte(data))
+	Expect(err).To(BeNil())
+
+	keyPEMBlock, err := getPrivateKey([]byte(data))
+	Expect(err).To(BeNil())
+
+	certs, err := X509KeyPair(certPEMBlock, keyPEMBlock, []byte{})
+	Expect(err).To(BeNil())
+	Expect(certs.Certificate).ToNot(BeNil())
+	Expect(certs.PrivateKey).ToNot(BeNil())
+}
+
+func X509KeyPair(certPEMBlock, keyPEMBlock, pw []byte) (cert tls.Certificate, err error) {
+	var certDERBlock *pem.Block
+	for {
+		certDERBlock, certPEMBlock = pem.Decode(certPEMBlock)
+		if certDERBlock == nil {
+			break
+		}
+		if certDERBlock.Type == "CERTIFICATE" {
+			cert.Certificate = append(cert.Certificate, certDERBlock.Bytes)
+		}
+	}
+
+	if len(cert.Certificate) == 0 {
+		err = errors.New("crypto/tls: failed to parse certificate PEM data")
+		return
```

Comment:
just curious with how Golang operates here... you're returning only the `return` keyword without the `err` yet for expected return types in the function declaration, you specify that there will be 2 objects that are returned `cert.Certificate, err`... does `return` interpret `err` to be the `err` output of this specific case? And where is it defined that a `cert` is returned as well?

Reply:
@bdlb77 So this is using the named return values - 

```
func X509KeyPair(certPEMBlock, keyPEMBlock, pw []byte) (cert tls.Certificate, err error) {
```

These are variables defined at the top of the function. The return statement without argument will return the named returned values.

Reply:
ah neat thank you!